container/src/state.rs

```diff
@@ -0,0 +1,218 @@
+
+const KAPS_ROOT_PATH: &str = "/var/run/kaps/containers";
+const OCI_VERSION: &str = "0.2.0";
+const STATE_FILE: &str = "state.json";
+
+/// Container runtime status
+#[derive(Serialize, Deserialize, Debug, PartialEq, Copy, Clone)]
+pub enum Status {
+    #[serde(rename = "creating")]
+    Creating,
+    #[serde(rename = "created")]
+    Created,
+    #[serde(rename = "running")]
+    Running,
+    #[serde(rename = "stopped")]
+    Stopped,
+}
+
+impl Default for Status {
+    fn default() -> Self {
+        Status::Creating
+    }
+}
+
+/// Represent the state of the running container.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ContainerState {
+    id: String,
+    /// OCI version.
+    oci_version: String,
+    /// Runtime state of the container.
+    pub status: Arc<RwLock<Status>>,
+    /// ID of the container process.
+    pub pid: i32,
+    /// Path to the bundle.
+    bundle: PathBuf,
+}
+
+impl Default for ContainerState {
+    fn default() -> Self {
+        ContainerState {
+            oci_version: OCI_VERSION.to_string(),
+            id: String::default(),
+            status: Arc::new(RwLock::new(Status::default())),
+            pid: 0,
+            bundle: PathBuf::default(),
+        }
+    }
+}
+
+impl ContainerState {
+    pub fn new(id: &str, bundle_path: &str) -> Result<Self> {
+        ContainerState::_new(id, bundle_path, KAPS_ROOT_PATH)
+    }
+
+    fn _new(id: &str, bundle_path: &str, container_dir: &str) -> Result<Self> {
+        let bundle = PathBuf::from(bundle_path);
+        let container_path = PathBuf::from(container_dir).join(id);
+
+        if container_path.as_path().exists() {
+            return Err(Error::ContainerExists(format!(
+                "A container with the id '{}' already exists",
+                id
+            )));
+        }
+
+        // create the container directory
+        fs::create_dir_all(&container_path).map_err(Error::CreateStateFile)?;
+
+        // create the `state.json` file of the container
+        File::create(container_path.join(STATE_FILE)).map_err(Error::CreateStateFile)?;
+
+        let container_state = ContainerState {
+            id: id.to_string(),
+            bundle,
+            ..Default::default()
+        };
+
+        container_state._save(container_dir)?;
+
+        Ok(container_state)
+    }
+
+    /// Get the current runtime status of the container.
+    ///
+    /// As The container status is a RwLock,
+    /// calling this function results in acquiring a read lock on the status.
+    fn _status(&self) -> Result<Status> {
+        let container_status = Arc::clone(&self.status);
+        let container_status = container_status
+            .read()
+            .map_err(|e| Error::StatusLockPoisoned(e.to_string()))?;
+
+        Ok(*container_status)
+    }
+
+    /// Save the container state.
+    ///
+    /// The container state file must already have been created.
+    fn _save(&self, container_dir: &str) -> Result<()> {
```

Comment:
This does not need to be `_` prefixed. a `_foo` function typically means it's a private function that's wrapped into a public one called `foo`. Here we only have `_save`, so you can just call it `save`.